### PR TITLE
Don't Panic on Missing Dirs

### PR DESCRIPTION
**Pull request by rothrock**
Issue: 1753
Figure out what's missing, create it, and keep going.

### CODE REVIEW

File: cmd/influxd/run.go [hunk 1]
```diff
@@ -155,8 +156,11 @@ func Run(config *Config, join, version string, logWriter *os.File) (*messaging.B
 
 	// unless disabled, start the loop to report anonymous usage stats every 24h
 	if !config.ReportingDisabled {
-		clusterID := b.Broker.Log().Config().ClusterID
-		go s.StartReportingLoop(version, clusterID)
+		// Make sure we have a config object b4 we try to use it.
+		if configObj := b.Broker.Log().Config(); configObj != nil {
```
Comment (otoolep) [inline cmd/influxd/run.go]: How could this happen? The `nil`?

File: cmd/influxd/run.go [hunk 2]
```diff
@@ -297,12 +292,14 @@ func openServer(config *Config, b *influxdb.Broker, initializing, configExists b
...
 
 // initializes a new server that does not yet have an ID.
-func initializeServer(u *url.URL, s *influxdb.Server, b *influxdb.Broker, w io.Writer) {
+func initializeServer(u *url.URL, s *influxdb.Server, b *influxdb.Broker, w io.Writer, initBroker bool) {
 	// TODO: Create replica using the messaging client.
 
-	// Create replica on broker.
-	if err := b.CreateReplica(1, u); err != nil {
-		log.Fatalf("replica creation error: %s", err)
+	if initBroker {
```
Comment (otoolep) [inline cmd/influxd/run.go]: A server needs to create replicas on the broker regardless of whether this node is initializing as a broker. In other words, this won't work if the node is running as a server *only*.

File: cmd/influxd/run.go (hunk 1 above)
Comment (rothrock) [inline cmd/influxd/run.go]: Well, it's what was causing the panics initially seen.

Comment (rothrock) [inline cmd/influxd/run.go]: Now, I think I've defended this from ever actually happening.

**Comment by otoolep** [comment]
Couple of things -- I think the design of checking if directories exist to determine the role of a node (server, data, or both) is broken. I think we should have some explicit state somewhere storing this (and derived from the "role" if it doesn't exist). Then check if the required directories exist, according to the state. I believe @toddboom agrees with me.

Also, this change seems to be only valid if a node is *both* a broker and a server. This is not necessarily the case.

File: cmd/influxd/run.go (hunk 2 above)
Comment (rothrock) [inline cmd/influxd/run.go]: Does every node need to have a raft directory?

Comment (otoolep) [inline cmd/influxd/run.go]: No, only Broker nodes.

**Comment by rothrock** [comment]
Oh sure, I think just checking for directory existence isn't a very strong check of viability. Nevertheless, this is what the previous code did. This change doesn't address node-only or broker-only configurations. The run function just tries to open a broker and to start a server. I haven't changed any of that logic, I don't think.

**Comment by otoolep** [comment]
OK, @rothrock and I spoke about this. I wanted to be clear that there is no issue unless the directory tree is in an inconsistent state. This could only happen (that I know of) if a Raft directory was deleted, and not the data directory, or vice-versa. Re-creating the directories should fix the issue, but I'm not sure it's what the user wanted since they may have just attempted to blow *everything* away. Not sure. 

I agree this code definitely needs work though -- it's got rather complex. It will be overhauled when "role" support comes back in.

**Comment by otoolep** [comment]
Perhaps we could add logging to the output when the directories are recreated? That way at least we warn the user that we have found the tree in an inconsistent state, and attempting to repair.

**Comment by rothrock** [comment]
Sure, I can say "...not found, creating..."

**Comment by otoolep** [comment]
OK, thanks. +1 from me.

**Pull request merged**